### PR TITLE
Push new STHs through ClusterNodeState

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/log/tree_signer_test.cc [hunk 1]
```diff
@@ -222,19 +222,17 @@ TYPED_TEST(TreeSignerTest, ResumeClean) {
...
 
   EXPECT_EQ(TS::OK, this->tree_signer_->UpdateTree());
-  SignedTreeHead sth;
 
-  EXPECT_EQ(DB::LOOKUP_OK, this->db()->LatestTreeHead(&sth));
+  const SignedTreeHead sth(this->tree_signer_->LatestSTH());
 
   TS* signer2 = this->GetSimilar();
-  EXPECT_EQ(signer2->LastUpdateTime(), sth.timestamp());
+  //  EXPECT_EQ(signer2->LastUpdateTime(), sth.timestamp());
```
Comment (pphaneuf) [inline cpp/log/tree_signer_test.cc]: Un-comment or delete?

**Comment by pphaneuf** [comment]
Could you rebase on top of the latest super_duper?

File: cpp/log/cluster_state_controller-inl.h [hunk 2]
```diff
@@ -65,6 +68,11 @@ void ClusterStateController<Logged>::NewTreeHead(
     CHECK_GE(sth.timestamp(), local_node_state_.newest_sth().timestamp());
   }
   local_node_state_.mutable_newest_sth()->CopyFrom(sth);
+  // We must, by definition, have a contiguous tree size of at least what's
+  // covered by the new STH:
+  if (sth.tree_size() > local_node_state_.contiguous_tree_size()) {
```
Comment (pphaneuf) [inline cpp/log/cluster_state_controller-inl.h]: Good catch, but I expect we'll be blowing away ```contiguous_tree_size``` shortly... :smiley:

File: cpp/log/cluster_state_controller-inl.h [hunk 3]
```diff
@@ -121,7 +129,7 @@ void ClusterStateController<Logged>::PushLocalNodeState(
 
   const util::Status status(store_->SetClusterNodeState(local_node_state_));
   if (!status.ok()) {
-    LOG(WARNING) << status;
+    LOG(WARNING) << "Couldn't set ClusterNodeState: " << status;
```
Comment (pphaneuf) [inline cpp/log/cluster_state_controller-inl.h]: Just thought of it now, but we might have a good number of spots like this where we could put ```LOG_IF``` to good profit!

File: cpp/log/cluster_state_controller-inl.h [hunk 4]
```diff
@@ -170,11 +180,62 @@ void ClusterStateController<Logged>::OnServingSthUpdated(
     LOG(WARNING) << "Cluster has no Serving STH!";
     actual_serving_sth_.reset();
   } else {
+    // TODO(alcutter): Validate STH and verify consistency with whatever we've
+    // already got locally.
+    if (!update.handle_.Entry().has_timestamp() ||
+        update.handle_.Entry().timestamp() == 0) {
```
Comment (pphaneuf) [inline cpp/log/cluster_state_controller-inl.h]: Did you get that in some case?

When ```has_timestamp()``` is false, ```timestamp()``` returns 0, BTW.

File: cpp/log/cluster_state_controller-inl.h (hunk 3 above)
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: Good idea.
I've done this one, and we can roll the others over as we spot them.

File: cpp/log/cluster_state_controller-inl.h (hunk 4 above)
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: Ah, just off the dodgy half-hearted initialisation code we had which shoved in an empty file for `serving_sth`

File: cpp/log/cluster_state_controller-inl.h (hunk 2 above)
Comment (AlCutter) [inline cpp/log/cluster_state_controller-inl.h]: Yup, coming soon to a PR near you...

File: cpp/log/tree_signer-inl.h [hunk 5]
```diff
@@ -201,46 +199,55 @@ util::Status TreeSigner<Logged>::SequenceNewEntries() {
...
-  VLOG(1) << "Building tree";
+
+  // Ensure we're in a position to be able to fetch entries from etcd.
+  const util::StatusOr<ct::SignedTreeHead> serving_sth(
+      consistent_store_->GetServingSTH());
+  if (!serving_sth.ok()) {
+    // TODO(alcutter): Should we allow this?
+    //                 Is this really a log provisioning failure or corrupted
+    //                 consistent store?
+    LOG(WARNING) << "Couldn't get current serving STH: "
```
Comment (pphaneuf) [inline cpp/log/tree_signer-inl.h]: Could this fit on a single line, or is GitHub playing tricks on me?

**Comment by pphaneuf** [comment]
LGTM

**Pull request merged**